apps/mobile/src/components/browser/approver-sheet/utils.ts

```diff
@@ -0,0 +1,88 @@
+import {
+  deriveBitcoinAccountFromAddress,
+  useBitcoinAccounts,
+} from '@/store/keychains/bitcoin/bitcoin-keychains.read';
+import { makeAccountIdentifer } from '@/store/utils';
+import { bytesToHex } from '@noble/hashes/utils';
+
+import {
+  getBitcoinInputAddress,
+  getBtcSignerLibNetworkConfigByMode,
+  getPsbtAsTransaction,
+  getPsbtTxInputs,
+  inferPaymentTypeFromAddress,
+  payerToBip32Derivation,
+  payerToTapBip32Derivation,
+} from '@leather.io/bitcoin';
+import {
+  extractAccountIndexFromDescriptor,
+  extractFingerprintFromDescriptor,
+} from '@leather.io/crypto';
+import { BitcoinNetworkModes } from '@leather.io/models';
+import { RpcRequests } from '@leather.io/rpc';
+
+export function normalizeSignAtIndex(signAtIndex: number | number[] | undefined) {
+  if (Array.isArray(signAtIndex)) {
+    return signAtIndex;
+  }
+  if (signAtIndex === undefined) {
+    return undefined;
+  }
+  return [signAtIndex];
+}
+
+interface InfuseBip32Props {
+  psbtHex: string;
+  networkMode: BitcoinNetworkModes;
+  bitcoinAccounts: ReturnType<typeof useBitcoinAccounts>['list'];
+  accountId: string;
+  signAtIndex: number | number[] | undefined;
+}
+
+export function infuseBip32({
+  psbtHex,
+  networkMode,
+  bitcoinAccounts,
+  accountId,
+  signAtIndex: _signAtIndex,
+}: InfuseBip32Props) {
+  const signAtIndex = normalizeSignAtIndex(_signAtIndex) ?? [];
+  const tx = getPsbtAsTransaction(psbtHex);
+
+  const inputs = getPsbtTxInputs(tx);
+  inputs.forEach((input, idx) => {
+    if (input.bip32Derivation || input.tapBip32Derivation) return;
+    const bitcoinNetwork = getBtcSignerLibNetworkConfigByMode(networkMode);
+    const addr = getBitcoinInputAddress(input, bitcoinNetwork);
+    const paymentType = inferPaymentTypeFromAddress(addr);
+    const bitcoinAccount = deriveBitcoinAccountFromAddress(bitcoinAccounts, addr);
+    if (bitcoinAccount?.descriptor) {
+      const accountIdx = extractAccountIndexFromDescriptor(bitcoinAccount?.descriptor);
+      const accountFingerprint = extractFingerprintFromDescriptor(bitcoinAccount?.descriptor ?? '');
+
+      const accAccountId = makeAccountIdentifer(accountFingerprint, accountIdx);
+      if (accAccountId === accountId) {
+        signAtIndex.push(idx);
+      } else if (accAccountId !== accountId && signAtIndex.includes(idx)) {
+        throw new Error('PSBT asks to sign an input that is not ours');
+      }
+    }
+
+    const payer = bitcoinAccount?.derivePayer({ addressIndex: 0 });
+    if (payer) {
+      if (paymentType === 'p2tr') {
+        tx.updateInput(idx, {
```

Comment:
Doesn't his try and update all the inputs within a psbt? if a psbt has 3 inputs, and the user passes `signAtIndex: [0,1]` then we only want to adjust 0 and 1, and this would do the same for 2 no?

Reply:
I think it's important that we only modify the inputs we're asked to. This is an easy fix, right? We just iterate `signAtInputs` instead?